File: day11/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}
	file, _ := os.Open(filename)
	defer file.Close()

	// Build adjacency list
	graph := make(map[string][]string)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ": ")
		node := parts[0]
		outputs := strings.Fields(parts[1])
		graph[node] = outputs
	}

	// Count paths from "you" to "out" using memoization
	memo := make(map[string]int)

	var countPaths func(node string) int
	countPaths = func(node string) int {
		if node == "out" {
			return 1
		}

		if val, exists := memo[node]; exists {
			return val
		}

		total := 0
		for _, next := range graph[node] {
			total += countPaths(next)
		}

		memo[node] = total
		return total
	}

	fmt.Println(countPaths("you"))
}

File: day11/haskell/part1.hs
import System.Environment (getArgs)
import qualified Data.Map as Map
import Data.List (foldl')

parseLine :: String -> (String, [String])
parseLine line = (node, outputs)
  where
    (node, rest) = break (== ':') line
    outputs = words $ drop 2 rest

-- Count paths using memoization
countPaths :: Map.Map String [String] -> String -> Map.Map String Int -> (Int, Map.Map String Int)
countPaths _ "out" memo = (1, memo)
countPaths graph node memo =
  case Map.lookup node memo of
    Just val -> (val, memo)
    Nothing ->
      let outputs = Map.findWithDefault [] node graph
          (total, memo') = foldl' addPath (0, memo) outputs
      in (total, Map.insert node total memo')
  where
    addPath (acc, m) next =
      let (count, m') = countPaths graph next m
      in (acc + count, m')

main :: IO ()
main = do
  args <- getArgs
  let filename = if null args then "input.txt" else head args
  content <- readFile filename
  let graph = Map.fromList $ map parseLine $ lines content
      (result, _) = countPaths graph "you" Map.empty
  print result
